Sources/ProcedureObserver.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  ProcedureKit
+//
+//  Copyright © 2016 ProcedureKit. All rights reserved.
+//
+
+import Foundation
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses to receive
+ events at state transitions.
+ */
+public protocol ProcedureObserver {
+
+    /**
+     Observer gets notified when it is attached to a procedure.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func didAttachTo(procedure: Procedure)
+}
+
+public extension ProcedureObserver {
+
+    /**
+     Default implementation does nothing.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func didAttachTo(procedure: Procedure) { }
+}
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses. They 
+ will receive a callback when the operation starts.
+ */
+public protocol WillExecuteProcedureObserver: ProcedureObserver {
+
+    /**
+     The procedure will execute.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func willExecute(procedure: Procedure)
+}
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses. They 
+ will receive a callback when the operation cancels.
+ */
+public protocol WillCancelProcedureObserver: ProcedureObserver {
+
+    /**
+     The procedure will cancel.
+
+     - parameter procedure: the observed `Procedure`.
+     - parameter errors: an array of [Error] types.
+     */
+    func willCancel(procedure: Procedure, errors: [Error])
+}
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses. They 
+ will receive a callback when the operation cancels.
+ */
+public protocol DidCancelProcedureObserver: ProcedureObserver {
+
+    /**
+     The procedure did cancel.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func didCancel(procedure: Procedure)
+}
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses. They 
+ will receive a callback when the Procedure produces another operation.
+ */
+public protocol DidProduceOperationProcedureObserver: ProcedureObserver {
```

Comment:
Yeah, I thought about that. Maybe.

But here, the procedure has just produced an (NS)Operation. Not necessarily a procedure. So thought it sounded misleading.